apps/api/src/services/programs.ts

```diff
@@ -0,0 +1,110 @@
+import type {
+  majorRequirementsQuerySchema,
+  minorRequirementsQuerySchema,
+  specializationRequirementsQuerySchema,
+} from "$schema";
+import type { database } from "@packages/db";
+import { eq, sql } from "@packages/db/drizzle";
+import { degree, major, minor, specialization } from "@packages/db/schema";
+import type { z } from "zod";
+
+export class ProgramsService {
+  constructor(private readonly db: ReturnType<typeof database>) {}
+
+  async getMajors() {
+    const majorSpecialization = this.db.$with("major_specialization").as(
+      this.db
+        .select({
+          id: major.id,
+          name: major.name,
+          specializations: sql`array_agg(${specialization.id})`.as("specializations"),
+        })
+        .from(major)
+        .innerJoin(specialization, eq(major.id, specialization.majorId))
+        .groupBy(major.id),
+    );
+
+    return this.db
+      .with(majorSpecialization)
+      .select({
+        id: majorSpecialization.id,
+        name: majorSpecialization.name,
+        specializations: majorSpecialization.specializations,
+        type: degree.name,
+        division: degree.division,
+      })
+      .from(majorSpecialization)
+      .innerJoin(major, eq(majorSpecialization.id, major.id))
+      .innerJoin(degree, eq(major.degreeId, degree.id));
+  }
+
+  async getMinors() {
+    return this.db
+      .select({
+        id: minor.id,
+        name: minor.name,
+      })
+      .from(minor);
+  }
+
+  async getSpecializations() {
+    return this.db
+      .select({
+        id: specialization.id,
+        majorId: specialization.majorId,
+        name: specialization.name,
+      })
+      .from(specialization);
+  }
+
+  async getMajorRequirements(query: z.infer<typeof majorRequirementsQuerySchema>) {
+    return await this.getProgramRequirements("major", query);
+  }
+
+  async getMinorRequirements(query: z.infer<typeof minorRequirementsQuerySchema>) {
+    return await this.getProgramRequirements("minor", query);
+  }
+
+  async getSpecializationRequirements(
+    query: z.infer<typeof specializationRequirementsQuerySchema>,
+  ) {
+    return await this.getProgramRequirements("specialization", query);
+  }
+
+  private async getProgramRequirements(
+    programType: "major" | "minor" | "specialization",
+    query:
+      | z.infer<typeof majorRequirementsQuerySchema>
+      | z.infer<typeof minorRequirementsQuerySchema>
+      | z.infer<typeof specializationRequirementsQuerySchema>,
+  ) {
+    switch (programType) {
+      case "major":
+        return await this.getProgramRequirementsInner(major, query);
+      case "minor":
+        return await this.getProgramRequirementsInner(minor, query);
+      case "specialization":
+        return await this.getProgramRequirementsInner(specialization, query);
+    }
+  }
```

Comment:
Interesting way of implementing this and I think on the more verbose side. Why not have `getProgramRequirements` take a tagged union of `programType` literals and `query` schemas?

Reply:
fair point